src/gravel/controllers/nodes/local.py

```diff
@@ -0,0 +1,185 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+import asyncio
+from enum import Enum
+from typing import List
+
+import psutil
+from pydantic import BaseModel, Field
+
+# Minimum requirements for individual host validation:
+# NOTE(jhesketh): These are obviously hardcoded for the moment, but may be
+#                 better moved into some kind of support matrix (eg feature
+#                 vs hardware etc).
+#                 These are also currently somewhat arbitrary and will need
+#                 tuning.
+AQUARIUM_MIN_CPU = 2
+AQUARIUM_MIN_MEMORY_GB = 2
+AQUARIUM_MIN_ROOT_DISK_GB = 10
+
+
+class HardwareQualifiedEnum(int, Enum):
+    QUALIFIED = 0
+    ERROR = 1
+
+
+class CPUQualifiedModel(BaseModel):
+    qualified: bool = Field("The CPU is sufficient")
+    min_cpu: int = Field("Minimum number of CPU's")
+    actual_cpu: int = Field("Actual number of CPU's")
+    error: str = Field("CPU didn't meet requirements")
+    status: HardwareQualifiedEnum = Field(HardwareQualifiedEnum.QUALIFIED)
+
+
+class MemoryQualifiedModel(BaseModel):
+    qualified: bool = Field("The memory is sufficient")
+    min_mem: int = Field("Minimum amount of memory (GB)")
+    actual_mem: int = Field("Actual amount of memory (GB)")
+    error: str = Field("Memory didn't meet requirements")
+    status: HardwareQualifiedEnum = Field(HardwareQualifiedEnum.QUALIFIED)
+
+
+class RootDiskQualifiedModel(BaseModel):
+    qualified: bool = Field("The root disk size is sufficient")
+    min_disk: int = Field("Minimum size of root disk (GB)")
+    actual_disk: int = Field("Actual size of root disk (GB)")
+    error: str = Field("Root disk didn't meet requirements")
+    status: HardwareQualifiedEnum = Field(HardwareQualifiedEnum.QUALIFIED)
+
+
+class LocalhostQualifiedModel(BaseModel):
+    all_qualified: bool = Field("The localhost passes validation")
+    cpu_qualified: CPUQualifiedModel = Field("CPU qualification details")
+    mem_qualified: MemoryQualifiedModel = Field("Memory qualification details")
+    root_disk_qualified: RootDiskQualifiedModel = Field(
+        "Root disk qualification details"
+    )
+
+
+async def validate_cpu() -> CPUQualifiedModel:
+    """
+    Validates the localhost meets the minium CPU requirements.
+    """
+    qualified: bool = True
+    min_cpu: int = AQUARIUM_MIN_CPU
+    actual_cpu: int = psutil.cpu_count()
+    error: str = ""
+    status: HardwareQualifiedEnum = HardwareQualifiedEnum.QUALIFIED
+
+    if actual_cpu < min_cpu:
+        qualified = False
+        error = (
+            "The node does not have a sufficient number of CPU cores. "
+            "Required: %d, Actual: %d." % (min_cpu, actual_cpu)
+        )
+        status = HardwareQualifiedEnum.ERROR
```

Comment:
The problem with this is that it is not conveying what the error is to the consumer. As we discussed before, the frontend will not be using the error message as is. What it will need is an error code and information that allows the frontend to, depending on the error code, provide insight to the user.

I'd suggest having error enums for different scenarios, even if at the moment we only support `some_value < target_value`.

Reply:
Yep, very good point, thanks.